src/components/AdminPanelLayout/UsersPane.tsx

```diff
@@ -1,57 +1,157 @@
+
+  const dispatch = useDispatch();
+  const history = useHistory();
 
   useMount(async () => {
     setUsers(await getUserList());
     setLoading(false);
   });
 
+  const handleRoleChange = (role: Role, user: User) => {
+    if (user.id === props.currentUser.id && !warningShown) {
+      setShowWarning(true);
+      setUserToUpdate([role, user]);
+    }
+    else {
+      updateUser(role, user);
+    }
+  };
+
+  const updateUser = (role: Role, user: User) => {
+    user.role = role;
+    addUserToUpdate({
+      userId: user.id,
+      newRole: role
+    });
+  };
+
   const addUserToUpdate = (request: ChangeUserRoleRequest) => {
     const requestIndex = updateRequests.findIndex((req) => req.userId === request.userId);
     if (requestIndex < 0) {
       setUpdateRequests([...updateRequests, request]);
     }
     else {
-      setUpdateRequests(updateRequests.splice(requestIndex, 1, request));
+      const newRequests = [...updateRequests];
+      newRequests[requestIndex] = request;
+      setUpdateRequests(newRequests);
     }
   };
 
+  const updateCurrentUserInWebstate = (user: User) => {
+    dispatch(setCurrentUser(user));
+  };
+
   const commitUsers = async () => {
     try {
       const response = await (changeUserRole(updateRequests));
-      setCommitResponsecode(response.status);
+      setCommitResponseCode(response.status);
       setUpdateRequests([]);
+      if (warningShown) {
+        updateCurrentUserInWebstate(await getCurrentUserFromServer());
+        history.push('/');
+      }
     } catch (e) {
-      setCommitResponsecode((e as AxiosResponse<string>).status);
+      setCommitResponseCode((e as AxiosResponse).status);
+    }
+    window.setTimeout(() => setCommitResponseCode(undefined), 2000);
+  };
+
+  const ConfirmRoleChange = () => {
+
+    const closeModal = () => {
+      setShowWarning(false);
+    };
+
+    const handleCancel = () => {
+      closeModal();
+      setUserToUpdate(undefined);
+    };
+
+    const handleConfirm = () => {
+      closeModal();
+      setWarningShown(true);
+      if (userToUpdateOnConfirm) {
+        updateUser(userToUpdateOnConfirm[0], userToUpdateOnConfirm[1]);
+      }
+    };
+
+    return (
+      <Layout>
+        <Modal
+          show={showWarning}
+          backdrop={'static'}
+          keyboard={false}
+        >
+          <ModalHeader style={{color: '#fff', backgroundColor: '#d9534f'}}>
+            CAUTION
+          </ModalHeader>
+          <ModalBody>
+            You are editing your own role. Doing so could lead to unexpected behavior, and you may be denied access to the
+            admin panel. Are you certain you wish to proceed?
+          </ModalBody>
+          <ModalFooter>
+            <Button variant={'primary'} onClick={handleCancel}>Cancel</Button>
+            <Button variant={'secondary'} onClick={handleConfirm}>Confirm</Button>
+          </ModalFooter>
+        </Modal>
+      </Layout>
+    );
+  };
+
+  const CommitStatus = () => {
```

Comment:
@jasekiw Would you rather inline this component? I thought it was pretty messy-looking with that nested ternary operator so I separated it out into this component which seems a lot more readable to me. Let me know if you prefer it inline though and I can change it back.